File: dotnet/ClientFiltering/Models/FilterCriteria.cs
namespace ClientFiltering.Models;

[DataContract]
[JsonConverter(typeof(FilterCriteriaJsonConverter))]
public record FilterCriteria
{
    /// <summary>
    /// The field to operate on
    /// </summary>
    [DataMember]
    public required string FieldName { get; init; }

    /// <summary>
    /// The operator
    /// </summary>
    [DataMember]
    [JsonConverter(typeof(FlexibleEnumConverter<LogicalOperators>))]
    public LogicalOperators Op { get; init; } = LogicalOperators.And;

    /// <summary>
    /// The comparison operator. Serialized as <c>logicalOperator</c>.
    /// </summary>
    [DataMember]
    [JsonPropertyName("logicalOperator")]
    [JsonConverter(typeof(FlexibleEnumConverter<RelationalOperators>))]
    public RelationalOperators Relation { get; init; }

    /// <summary>
    /// The values to use for comparison
    /// </summary>
    [DataMember]
    public required IReadOnlyCollection<string?> Values { get; init; } = [];
}

File: javascript/src/lib/filtercriteria.ts
import { Operators, Logic } from "./enums";
export class FilterCriteria {
  /// <summary>
  /// The field to operate on
  /// </summary>
  public FieldName: string;
  /// <summary>
  /// The operator
  /// </summary>
  public Op = Operators.And;
  /// <summary>
  /// The logical operator for the function
  /// </summary>
  public LogicalOperator = Logic.Equals;
  /// <summary>
  /// The values to use for comparison
  /// </summary>
  public Values = Array.from<string | null>([]);
}
